### ISSUE TITLE
测试模型遇到的问题

### ISSUE THREAD

**Issue by tianboguangding** [AlfredXiangWu/face_verification_experiment#42]
吴老师，您好：
先感谢您分享的网络模型！
当测试模型时遇到两个问题：
1. 按照网络A_deploy，提取fc1层的特征应该是512维（按这个正确率是93.8%），但看保存特征的mat里，特征维数是256维，想问一下，您提取的是那一层特征？
2.在测试网络B_deploy时，出现 ：
...insert_splits.cpp:35] Unknown blob input label to layer 1 的错误，想问一下这个怎么解决？

十分感谢！

**Comment by AlfredXiangWu** [comment]
1. 特征层对于A是eltwise6，对于B是eltwise_fc1
2. 错误信息的字面意思，unknown blob input label to layer 1，你最好自己检查一下网络配置吧

**Comment by tianboguangding** [comment]
多谢

训练测试集我是把它分成9:1,而没有把所有数据拿去训练，结果按这个训练出的模型，在lfw上正确率只有65.8%，原因是什么？（孙祎的五点定位，检测，再按照您的方式alignment，最后是144*144的灰度图片，样本准备过程应该没有什么问题,lr：0.001，weight_decay:0.005）

**Comment by AlfredXiangWu** [comment]
提取fc1 512维的特征是93.8%的准确率, 而变成eltwise6的256维就只有65.8%？

**Comment by tianboguangding** [comment]
您的模型变成eltwise6的256维准确率是97.7%；65.8%是我自己按上边网络重新训练模型并在lfw的测试结果，不知道问题出在哪儿？

**Comment by AlfredXiangWu** [comment]
65.8%这个结果太低了，我这边网络的迭代4w次，validation只有30%~40%的时候在lfw也可以到90%左右。我觉得首先还是仔细检查下数据预处理吧，训练集和测试集预处理的方法是否一致，因为训练集裁出来的144x144的图像，测试集lfw应该是128x128的。另外看一下你网络在训练时的收敛状况，你这个65.8%结果的网络validation的准确率是多少呢？

**Comment by tianboguangding** [comment]
hi，吴老师：
是的，预处理昨天我又重新做了一遍，先港中文的批处理一遍，之后训练集是按照您的说明裁剪的144X144灰度图，lfw也是128X128的灰度图
训练集是全部的img，val集是从中选取了三万多张，下面是训练的部分收敛情况：
......
I0331 12:59:07.620589  4824 solver.cpp:214] Iteration 230900, loss = 2.82966
I0331 12:59:07.621589  4824 solver.cpp:229]     Train net output #0: accuracy = 0.515625
I0331 12:59:07.621589  4824 solver.cpp:486] Iteration 230900, lr = 0.001
I0331 13:00:46.470243  4824 solver.cpp:294] Iteration 231000, Testing net (#0)
I0331 13:01:48.563794  4824 solver.cpp:330] Test loss: 1.67944
I0331 13:01:48.564795  4824 solver.cpp:343]     Test net output #0: accuracy = 0.7663
I0331 13:01:48.940816  4824 solver.cpp:214] Iteration 231000, loss = 3.35271
I0331 13:01:48.940816  4824 solver.cpp:229]     Train net output #0: accuracy = 0.375
I0331 13:01:48.941817  4824 solver.cpp:486] Iteration 231000, lr = 0.001
I0331 13:03:28.806529  4824 solver.cpp:214] Iteration 231100, loss = 3.02206
I0331 13:03:28.806529  4824 solver.cpp:229]     Train net output #0: accuracy = 0.4375
I0331 13:03:28.807528  4824 solver.cpp:486] Iteration 231100, lr = 0.001
I0331 13:05:08.690242  4824 solver.cpp:214] Iteration 231200, loss = 2.77543
I0331 13:05:08.691241  4824 solver.cpp:229]     Train net output #0: accuracy = 0.484375
......
这个是不是不太正常？
而在lfw上的正确率还是很低。

**Comment by AlfredXiangWu** [comment]
我觉得你这个log变化很正常啊，然后你测试的时候lfw才只有65.8%？ 你检查下你提特征的代码吧，提特征的时候图像像素要除255归一化到[0,1]

**Comment by tianboguangding** [comment]
是的，测试lfw才只有65.8%，今天又重新试了两遍也都是67%或68%（迭代65万次时的模型）。
归一化因为用的是img=caffe.io.load_image(path,color = not image_as_grey)，所以img的图像之都是介于0~1之间。下面是我用的提特征的代码（根据您的caffe_ftr.py稍做整理）：

...........................................
# -*- coding: utf-8 -*-
import gzip
import zipfile
import cPickle
import time
import scipy.io as sio
import os

from collections import OrderedDict
import sklearn
import numpy as np
import matplotlib.pyplot as plt
import skimage.io
caffe_root='D:/caffe/caffe-windows-master02/bin/'
import sys
sys.path.insert(0,caffe_root + 'python')
import caffe
import sklearn.metrics.pairwise as pw

def extract_feature(img_list,network_proto_path,network_model_path,data_mean,layer_name,image_as_grey):

       net=caffe.Classifier(network_proto_path,network_model_path)
       caffe.set_mode_gpu()
       print 'net input:',net.inputs
       print 'net output:',net.outputs

       if data_mean is None:
              data_mean=np.zeros(1)
       net.transformer.set_mean('data',data_mean)

       if not image_as_grey:
              net.transformer.set_channel_swap('data',(2,1,0))

       net.transformer.set_input_scale('data',1)

       blobs=OrderedDict([(k,v.data) for k,v in net.blobs.items()])
       shp=blobs[layer_name].shape
       print blobs['data'].shape

       batch_size=blobs['data'].shape[0]
       print blobs[layer_name].shape

       if len(shp) is 2:
              features_shape=(len(img_list),shp[1])
       elif len(shp) is 3:
              features_shape=(len(img_list),shp[1],shp[2])
       elif len(shp) is 4:
              features_shape=(len(img_list),shp[1],shp[2],shp[3])

       features=np.empty(features_shape,dtype='float32',order='C')
       img_batch=[]
       for cnt, path in zip(range(features_shape[0]), img_list):
              img=caffe.io.load_image(path,color = not image_as_grey)#？？
              if image_as_grey and img.shape[2]!=1:
                     img=skimage.color.rgb2gray(img)
                     img=img[:,:,np.newaxis]
              if cnt ==0:
                     print 'image shape: ' , img.shape

              img_batch.append(img)

              if (len(img_batch)==batch_size) or cnt==features_shape[0]-1:#img_batch=20，所以...
                     scores=net.predict(img_batch,oversample=False)

                     blobs=OrderedDict([(k,v.data) for k,v in net.blobs.items()])
                     print '%d images processed' %(cnt+1,)#

                     #features[cnt-len(img_batch)+1:cnt,:,:,:]=blobs['fc1'][0:len(img_batch)-1,:,:,:].copy()#版本原因
                     features[cnt-len(img_batch)+1:cnt+1]=blobs[layer_name][0:len(img_batch)].copy()
                     img_batch=[]
       features=np.asarray(features,dtype='float32')
       return features

def labels_list_to_float(labels):
       int_labels = []
       for e in labels:
              try:
                     inte=int(e)
              except ValueError:
                     print '标签不是整数'
                     break
              int_labels.append(inte)
       if len(int_labels)==len(labels):
              return int_labels


       labels_unique = list(sorted(set(labels)))
       print labels[0:10]
       print labels_unique[0:10]

       dic = dict([(lb, i) for i, lb in zip(range(len(labels_unique)),labels_unique)])
       labels_float = [dic[a] for a in labels]

       return labels_float

def string_list_to_cells(labels):
       cells=np.ndarray(len(labels),dtype='object')
       for i in range(len(labels)):
              cells[i]=labels[i]
       return cells

def evaluate():
       file_list=open('D:/caffe/0pair_program/image_path.txt','r')#图像路径，不包括标签
       img_list=[]
       labels=[]
       for line in file_list:
              items=line.split()
              img_list.append(items[0].strip())
              
       label_list=open('D:/caffe/0pair_program/labels.txt','r')#相应标签
       for line in label_list:
              items=line.split()
              labels.append(items[0].strip())

       float_labels=labels_list_to_float(labels)#

       data_mean=None#无均值
       image_as_grey=True#灰度图像
       save_path='D:/caffe/face_verification_experiment/lfw_200000_all_64_normImg.mat'#生成并保存
       network_proto_path='D:/caffe/0pair_program/face_verification_experiment/DeepFace_set003.prototxt'#A
       network_model_path='D:/caffe/0pair_program/face_verification_experiment/DeepFace_set003_net_iter_200000_all_64.caffemodel'#迭代20万次时的模型
       layer_name='eltwise6'#A网络
       #layer_name='eltwise_fc1'#B网络

       ftr=extract_feature(img_list,network_proto_path,network_model_path,data_mean,layer_name,image_as_grey)
       
       if ftr.ndim==4 and ftr.shape[3]==1 and ftr.shape[2]==1:
              ftr=ftr[:,:,0,0]

       dic={'features':ftr,
            'labels':float_labels,
            'labels_original':string_list_to_cells(labels),
            'image_path':string_list_to_cells(img_list)}
       
       sio.savemat(save_path,dic)
       p=1
if __name__=='__main__':      
       evaluate()

........................................
因为caffe版本不同，所以用了您改过的caffe_ftr.py，又稍微调整一下，想请您看看是不是有什么问题。
十分感谢！

**Comment by tianboguangding** [comment]
hi，吴老师：
用我提特征的代码载入您的model从而生成的lfw.mat，检测正确率是96.8%，所以感觉提特征代码好像没什么问题；但载入我自己训练好的模型不是生成lfw.mat失败（Cannot copy param 0 weights from layer 'conv1'; shape mismatch.  Source param shape is 96 3 9 9 (23328); target param shape is 96 1 9 9 (7776). To learn this layer's parameters from scratch  rather than copying from a saved net, rename the layer）就是生成了正确率很低（65.8%），而之前您也说我训练的log变化是正常的，那这个问题是出现在哪里呢？

**Comment by AlfredXiangWu** [comment]
> Cannot copy param 0 weights from layer 'conv1'; shape mismatch. Source param shape is 96 3 9 9 (23328); target param shape is 96 1 9 9 (7776). To learn this layer's parameters from scratch rather than copying from a saved net, rename the layer

这个错误就是字面意思，彩色图像和灰度图像的问题。。。。
65.8%到底是什么问题我也不是特别确定，看log我觉得没问题。所以我还是觉得你提特征的部分有问题，尤其是你说你的代码有时候提取不成功，有时候提出来特征准确率是65.8%，这么不稳定的代码明显里面是存在bug的

**Comment by tianboguangding** [comment]
hi，吴老师：
我又重新做了一遍数据前处理加训练（c++训练，python提特征），有点郁闷，当用我代码载入您的模型(如LightenedCNN_A.caffemodel)提取包含特征的lfw.mat，检测其正确率仍是96.8%，但当载入我自己模型从而生成的plfw.mat,正确率却还是很低，总感觉不应该啊，如果提特征代码有问题，应该两个模型的所生成的lfw.mat正确率都很低，或者都是错误的。但如果生成的模型有问题那您也不会说训练log没问题。
对比了一下载入您的模型和我训练模型并从而生成的lfw.mat和plfw.mat里的feature，发现您的特征值大小在正负几十区间（24.04，-30.38，6.05等）绝对值普遍大，而我提的特征值一般是个位数区间（ 3.61，-4.32，0.23等）绝对值普遍小，这是为什么啊

**Comment by tianboguangding** [comment]
Hi,吴老师：
sorry，是您说的彩色图像和灰度图像的问题。现在好了，迭代12万次时，lfw测试准确率有94%。
不好意思，谢谢了。
